src/napari_graph/directed_graph.py

```diff
@@ -0,0 +1,418 @@
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from numba import njit, typed
+from numpy.typing import ArrayLike
+
+from napari_graph._base_graph import (
+    _EDGE_EMPTY_PTR,
+    BaseGraph,
+    _iterate_edges,
+    _remove_edge,
+)
+from napari_graph.undirected_graph import _UN_EDGE_SIZE
+
+"""
+Directed edge constants for accessing the directed graph buffer data.
+Each edge occupies _DI_EDGE_SIZE spaces on the graph buffer.
+_LL_DI_EDGE_POS indicates the displacement between the edge initial index and
+the **target** edge directed linked list position.
+
+Example of a directed graph edge buffer:
+[
+    source_node_buffer_id_0,
+    target_node_buffer_id_0,
+    source_edge_linked_list_0,
+    target_edge_linked_List_0,
+    source_node_buffer_id_1,
+    target_node_buffer_id_1,
+    source_edge_linked_list_1,
+    target_edge_linked_List_1,
+    ...
+]
+"""
+_DI_EDGE_SIZE = _UN_EDGE_SIZE + 1
+_LL_DI_EDGE_POS = 2
+
+
+@njit
+def _add_directed_edge(
+    buffer: np.ndarray,
+    node2src_edge: np.ndarray,
+    node2tgt_edge: np.ndarray,
+    empty_idx: int,
+    src_node: int,
+    tgt_node: int,
+) -> int:
+    """Add a single directed edge to `buffer`.
+
+    This updates the `buffer`'s source and target linked list
+    and the nodes to edges mappings.
+
+    NOTE: see `_add_undirected_edge` docs for comment about cache misses.
+
+     Returns
```

Comment:
```suggestion
    Returns
```

It'd be good to add parameter descriptions too — as noted elsewhere, this can be very useful for someone coming in to the code later. (Even your later self 😉)